test/Microsoft.Azure.Devices.Edge.Test/EndToEnd.cs

```diff
@@ -0,0 +1,160 @@
+                {
+                    var iotHub = new IotHub(
+                        Context.Current.ConnectionString,
+                        Context.Current.EventHubEndpoint,
+                        proxy);
+
+                    EdgeDevice device = (await EdgeDevice.GetIdentityAsync(
+                        Context.Current.DeviceId,
+                        iotHub,
+                        token)).Expect(() => new Exception("Device should have already been created in setup fixture"));
+
+                    var builder = new EdgeConfigBuilder(device.Id);
+                    Context.Current.Registry.ForEach(
+                        r => builder.AddRegistryCredentials(r.address, r.username, r.password));
+                    builder.AddEdgeAgent(edgeAgent).WithProxy(proxy, Protocol.Amqp);
+                    builder.AddEdgeHub(edgeHub).WithProxy(proxy, Protocol.Amqp);
+                    builder.AddModule("tempSensor", tempSensor);
+                    await builder.Build().DeployAsync(iotHub, token);
+
+                    var hub = new EdgeModule("edgeHub", device.Id, iotHub);
+                    var sensor = new EdgeModule("tempSensor", device.Id, iotHub);
+                    await EdgeModule.WaitForStatusAsync(
+                        new[] { hub, sensor },
+                        EdgeModuleStatus.Running,
+                        token);
+                    await sensor.WaitForEventsReceivedAsync(token);
+
+                    var sensorTwin = new ModuleTwin(sensor.Id, device.Id, iotHub);
+                    await sensorTwin.UpdateDesiredPropertiesAsync(
+                        new
+                        {
+                            properties = new
+                            {
+                                desired = new
+                                {
+                                    SendData = true,
+                                    SendInterval = 10
+                                }
+                            }
+                        },
+                        token);
+                    await sensorTwin.WaitForReportedPropertyUpdatesAsync(
+                        new
+                        {
+                            properties = new
+                            {
+                                reported = new
+                                {
+                                    SendData = true,
+                                    SendInterval = 10
+                                }
+                            }
+                        },
+                        token);
+                },
+                "Completed test '{Name}'",
+                name);
+        }
+
+        [Test]
+        public async Task ModuleToModuleDirectMethod(
+            [Values("Amqp", "Mqtt")] string edgeToCloud,
+            [Values("Amqp", "Mqtt")] string moduleToEdge)
+        {
+            string edgeAgent = Context.Current.EdgeAgent.Expect(() => new ArgumentException());
```

Comment:
you can use Precondition, isn't it?

Reply:
is this edgeAgent image full path? if so, please rename it as edgeAgentImage.

Reply:
> you can use Precondition, isn't it?

Precondition doesn't help here...I'm using `Expect` as a way to unwrap the value embedded inside the `Option`.

Reply:
> ...please rename it as edgeAgentImage...

That's been on my todo list; I'll do it now.

Reply:
Fixed.